ballerina/vector-store.bal

```diff
@@ -89,17 +89,135 @@ public distinct isolated class InMemoryVectorStore {
+    }
+
+    private isolated function entryMatchesFilters(VectorMatch|InMemoryVectorEntry entry, MetadataFilters filters) returns boolean|error {
+        Metadata? metadata = entry.chunk.metadata;
+        if metadata is () {
+            return false;
+        }
+        return check self.evaluateFilterNode(metadata, filters);
+    }
+
+    private isolated function evaluateFilterNode(Metadata content, MetadataFilters|MetadataFilter node) returns boolean|error {
+        if node is MetadataFilter {
+            return content.hasKey(node.key) ? self.compareValues(content.get(node.key), node.operator, node.value) : false;
+        }
+        boolean[] results = [];
+        foreach MetadataFilters|MetadataFilter child in node.filters {
+            boolean childResult = check self.evaluateFilterNode(content, child);
+            results.push(childResult);
+        }
+        return self.evaluateCondition(node.condition, results);
+    }
+
+    private isolated function evaluateCondition(MetadataFilterCondition condition, boolean[] results) returns boolean {
+        if condition == AND {
+            foreach boolean result in results {
+                if !result {
+                    return false;
+                }
+            }
+            return true;
+        }
+        foreach boolean result in results {
+            if result {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private isolated function compareValues(json left, MetadataFilterOperator operation, json right) returns boolean|error {
+        match operation {
+            EQUAL => {
+                return left == right;
+            }
+            NOT_EQUAL => {
+                return left != right;
+            }
+            IN => {
+                if right is json[] {
+                    foreach json value in right {
+                        if left == value {
+                            return true;
+                        }
+                    }
+                }
+                return false;
+            }
+            NOT_IN => {
+                if right is json[] {
+                    foreach json value in right {
+                        if left == value {
+                            return false;
+                        }
+                    }
+                    return true;
+                }
+                return false;
+            }
+            GREATER_THAN => {
+                return check left.cloneWithType(decimal) > check right.cloneWithType(decimal);
+            }
+            LESS_THAN => {
+                return check left.cloneWithType(decimal) < check right.cloneWithType(decimal);
+            }
+            GREATER_THAN_OR_EQUAL => {
+                return check left.cloneWithType(decimal) >= check right.cloneWithType(decimal);
+            }
+            LESS_THAN_OR_EQUAL => {
+                return check left.cloneWithType(decimal) <= check right.cloneWithType(decimal);
```

Comment:
Do we have a guarantee that these values are always decimal (or at for the time we don't care)?. Not all numeric values can be converted to decimal. For example float has nan, +inf and -inf and deciaml don't support those values (you will get an error)

Reply:
Had a discussion with @MohamedSabthar regarding this. Since this is a rare case, decided to keep the current changes